test/dsets.c

```diff
@@ -5681,6 +5681,306 @@ test_multiopen(hid_t file)
+    }
+#if H5_SIZEOF_LONG_DOUBLE != H5_SIZEOF_DOUBLE
+    /* long double */
+    {
+        long double orig_data[2][5] = {
+            {(long double)1.6081706885101836e+600L, (long double)-255.3209917099448032099170994480,
+             (long double)1.2677579992621376e-610L, (long double)64568.289448797700289448797700,
+             (long double)-1.0619721778839084e-750L},
+            {(long double)2.1499497833454840991499497833454840e+560L,
+             (long double)6.6562295504670740996562295504670740e-3,
+             (long double)-1.5747263393432150995747263393432150,
+             (long double)1.0711093225222612990711093225222612,
+             (long double)-9.8971679387636870998971679387636870e-1}};
+        long double new_data[2][5];
+
+        TESTING("    long double (setup)");
+
+        /* Define user-defined quad-precision floating-point type for dataset */
+        datatype  = H5Tcopy(H5T_NATIVE_LDOUBLE);
+        precision = 128;
+        if (H5Tset_precision(datatype, precision) < 0)
+            goto error;
+        if (H5Tset_fields(datatype, (size_t)127, (size_t)112, (size_t)15, (size_t)5, (size_t)107) < 0)
+            goto error;
+        offset = 5;
+        if (H5Tset_offset(datatype, offset) < 0)
+            goto error;
+        if (H5Tset_size(datatype, (size_t)16) < 0)
+            goto error;
+        if (H5Tset_ebias(datatype, (size_t)255) < 0)
+            goto error;
+
+        /* Create the data space */
+        if ((space = H5Screate_simple(2, size, NULL)) < 0)
+            goto error;
+
+        /* Create the dataset */
+        if ((dataset = H5Dcreate2(file, "long_double_type", datatype, space, H5P_DEFAULT, H5P_DEFAULT,
+                                  H5P_DEFAULT)) < 0)
+            goto error;
+
+        PASSED();
+
+        /*----------------------------------------------------------------------
+         * STEP 1: Test writing to it.
+         *----------------------------------------------------------------------
+         */
+        TESTING("    long double (write)");
+
+        if (H5Dwrite(dataset, H5T_NATIVE_LDOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, orig_data) < 0)
+            goto error;
+        PASSED();
+
+        /*----------------------------------------------------------------------
+         * STEP 2: Try to read the data we just wrote.
+         *----------------------------------------------------------------------
+         */
+        TESTING("    long double (read)");
+
+        /* Read the dataset back */
+        if (H5Dread(dataset, H5T_NATIVE_LDOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, new_data) < 0)
+            goto error;
+
+        /* Check that the values read are the same as the values written */
+        for (i = 0; i < (size_t)size[0]; i++) {
+            for (j = 0; j < (size_t)size[1]; j++) {
+                if (isnan(orig_data[i][j]))
+                    continue; /* skip if value is NaN */
+                if (!H5_LDBL_ABS_EQUAL(new_data[i][j], orig_data[i][j])) {
```

Comment:
Are we sure that is checking the complete precision? I would think it would fail if the complete precision were not initialized. Also, do we expect NaN values? I would think that would be an error.

Reply:
This just follows the other float and double parts. 
These would be issues that should be noted.

Reply:
These macros just get the difference of the two float, double, or long double values and then take the absolute value and compare it to the relevant epsilon value. It should be fine for checking the data.